Don't introduce spurious libLLVM dependency into julia binary

Keno opened the pull request:
This causes problems with the new RPATH handling since we now omit
lib/julia from the executable's RPATH. Fix this by not forcing
the executable to link against libLLVM, since we don't require it.

Fixes #17484.

Keno committed:
Don't introduce spurious libLLVM dependency into julia binary

This causes problems with the new RPATH handling since we now omit
lib/julia from the executable's RPATH. Fix this by not forcing
the executable to link against libLLVM, since we don't require it.

Fixes #17484.

vtjnash commented:
lgtm

Keno commented:
@tkelman Once this is merged we should trigger a rebuild of the nightlies.

tkelman commented:
essentially every commit to master triggers a new nightly, if it passes tests.

will this bring back the issue that https://github.com/JuliaLang/julia/pull/7696 was needed to fix? presumably this was here for a reason

Keno commented:
@JeffBezanson said the ui/Makefile change back then was optional so I don't think it'll be a problem.

tkelman commented:
optional in what sense?

vtjnash commented:
I think it's optional in the sense that it would allow us to hide errors in our `ld` invocation for building libjulia on some platforms (by delaying them onto we tried to link the files in ui)

tkelman commented:
linux travis failure looks real, undefined references to llvm when linking julia-debug. alternate approach?

Keno committed:
Provide link-time rpath

Unlike ld.so, ld does not understand `$ORIGIN`, so provide the full path to it at link
time to make sure it can find the appropriate dependencies.

Keno commented:
I pushed a fix.

vtjnash commented:
It seems the linux distros changed the linker to default reject this build configuration circa gcc 4.5? (source: http://fedoraproject.org/wiki/UnderstandingDSOLinkChange, https://wiki.ubuntu.com/NattyNarwhal/ToolchainTransition, others)

Keno commented:
No, I don't think that's the problem. We're not actually using those symbols in the executable, they're just undefined in libjulia-debug.so and the linker wants to make sure they exist somewhere (but for some reason it doesn't know about `$ORIGIN` - not sure why, the OS X linker doesn't have that problem).

tkelman commented:
Out of an abundance of caution, I'll build binaries from 9c9ad9f486f1f932254eda9453335d18e68ee003 for testing purposes. Search http://s3.amazonaws.com/julianightlies for the first 10 letters of that sha in a few hours, it'll probably beat Travis.

ranjanan commented:
I downloaded the Mac binary and got this libuv error on running Base.runtests():
```
Exception running test loading :
On worker 3:
LoadError: open: read-only file system (EROFS)
 in uv_error at ./libuv.jl:68 [inlined]
 in open at ./filesystem.jl:81
 in touch at ./file.jl:189
 in include_string at ./loading.jl:380
 in include_from_node1 at ./loading.jl:429
 in macro expansion at ./util.jl:180 [inlined]
 in runtests at /Volumes/Julia-0.5.0-dev-9c9ad9f486/Julia-0.5.app/Contents/Resources/julia/share/julia/test/testdefs.jl:7
 in #16 at /Volumes/Julia-0.5.0-dev-9c9ad9f486/Julia-0.5.app/Contents/Resources/julia/share/julia/test/runtests.jl:44
 in #501 at ./multi.jl:1193
 in run_work_thunk at ./multi.jl:844
 in macro expansion at ./multi.jl:1193 [inlined]
 in #500 at ./event.jl:46
while loading /Volumes/Julia-0.5.0-dev-9c9ad9f486/Julia-0.5.app/Contents/Resources/julia/share/julia/test/loading.jl, in expression starting on line 16
ERROR: LoadError: Some tests exited with errors.
 in (::##11#19)() at /Volumes/Julia-0.5.0-dev-9c9ad9f486/Julia-0.5.app/Contents/Resources/julia/share/julia/test/runtests.jl:72
 in cd(::##11#19, ::String) at ./file.jl:59
 in include_from_node1(::String) at ./loading.jl:426
 in process_options(::Base.JLOptions) at ./client.jl:262
 in _start() at ./client.jl:318
while loading /Volumes/Julia-0.5.0-dev-9c9ad9f486/Julia-0.5.app/Contents/Resources/julia/share/julia/test/runtests.jl, in expression starting on line 26
```

tkelman commented:
how did you install/mount it? I don't think that would be related?

ranjanan commented:
@tkelman I downloaded the dmg, extracted it and opened it. It opens up a terminal with julia running. I ran `versioninfo()` which worked. And then Base.runtests(). I admit that these two issues might be unrelated.

tkelman commented:
extracted with what commands? unless you move the contents to a local volume it will likely still be read only
